### ISSUE TITLE
Can you create authorization policies without using magic strings?

### ISSUE THREAD

**Issue by shaulbehr** [aspnet/Security#1689]
In [this article](https://docs.microsoft.com/en-us/aspnet/core/security/authorization/policies) I see that you declare your authorization policy using a magic string:
```
options.AddPolicy("AtLeast21", policy =>
    policy.Requirements.Add(new MinimumAgeRequirement(21)));
```
And then you tag your controller/endpoint with:
```
[Authorize(Policy = "AtLeast21")]
```
So, let's say I have multiple endpoints, with different age requirements for each. Do I then need to declare a separate policy for each age cutoff?
```
...
options.AddPolicy("AtLeast18", policy =>
    policy.Requirements.Add(new MinimumAgeRequirement(18)));
options.AddPolicy("AtLeast19", policy =>
    policy.Requirements.Add(new MinimumAgeRequirement(19)));
options.AddPolicy("AtLeast20", policy =>
    policy.Requirements.Add(new MinimumAgeRequirement(20)));
options.AddPolicy("AtLeast21", policy =>
    policy.Requirements.Add(new MinimumAgeRequirement(21)));
...
```
That's really not going to be the answer, is it...? I mean, I could obviously write a loop to add policies for every age between 1 and 120, but what happens if my AuthenticationRequirement has more than one variable? Are we going to write a quadruple-nested loop for my requirement that has four variables, so I can declare distinct policies for each one? And what if my policy takes a `double` as a parameter?

Please, tell me I've misunderstood how this works.

**Comment by mjrousos** [comment]
You can make the auto-generation of policies easier with a [custom IAuthorizationPolicyProvider](https://github.com/aspnet/Security/issues/670#issuecomment-174354109). I'll make a little sample and post it momentarily.

**Comment by shaulbehr** [comment]
@mjrousos I looked at the issue you linked above. Not clear how `IAuthorizationPolicyProvider` helps anything; it still looks like the policy is being declared as a string rather than a signature with variables. Look forward to seeing your code sample.

**Comment by Eilon** [comment]
@shaulbehr thanks for the great question, we will take a look at it ASAP and see what guidance we have for this.

**Comment by blowdart** [comment]
Generally what I've done in demos is have a static class with the strings as public static consts, then you use something like AuthorizationPolicies.Over21. That restricts the definition to one place.

Your example of AtLeast18, AtLeast19, AtLeast20, AtLeast21 is, well, a little contrived, could you perhaps provide one that's, well, more real world, where you end up with multiple policies that are pretty close to each other in terms of parameters and names?

**Comment by HaoK** [comment]
So you can autogen the policies based on the name too, where your policy provider parses the name and generates the right requirement instances based on that.

**Comment by mjrousos** [comment]
@shaulbehr, I've put a small sample of how you might handle this case with `IAuthorizationPolicyProvider` [here](https://github.com/mjrousos/IAuthorizationPolicyProviderSample).

What it boils down to is this: 

Authorization policies are only identified by their name string at the moment. So, I don't think you can completely get away from "AtLeast18", "AtLeast19", "AtLeast20", etc. in cases where you have 'parameterized' authorization policies. 

What you *can* do however (as shown in the sample) is hide that messiness. Using an `IAuthorizationPolicyProvider` allows you to dynamically generate all the AtLeast* policies you need without having to define them all individually. Similarly, I use a custom `AuthorizeAttribute` in the sample to allow developers to specify parameters (like age) when labeling APIs with [Authorize] attributes (those parameters are incorporated into policy name strings under the covers, but the developers never need to see that).

So, you don't get real parameterized auth policies, but it looks as if you do. :-) @Eilon, please let me know if you find any guidance with a better way to do this; I'm not aware of any, but would be interested to know if we have more official support for this sort of scenario.

**Comment by shaulbehr** [comment]
@mjrousos Thanks for that, I've copied your pattern in my project now. Can't say if it's working yet, since I have so many other things to convert before the project even compiles!

@blowdart I think your response is a little defensive. It doesn't matter that this particular example is contrived. The fact that parameter values have to be serialized into a string and deserialized in order to interpret the policy is, to my mind, a pretty serious design flaw. The old ASP.NET had this covered, so there's no reason why ASP.NET Core shouldn't be able to replicate the same behavior.

**Comment by PinpointTownes** [comment]
> Generally what I've done in demos is have a static class with the strings as public static consts, then you use something like AuthorizationPolicies.Over21.

"demos"... :trollface: 

> Your example of AtLeast18, AtLeast19, AtLeast20, AtLeast21 is, well, a little contrived, could you perhaps provide one that's, well, more real world, where you end up with multiple policies that are pretty close to each other in terms of parameters and names?

In your workshop "demos", you have a policy that restricts access to a (hardcoded) building based on badge requirements. One might want to adopt a similar system for room access, where the room number is not hardcoded.

Essentially, what's missing is a way to pass strongly typed data from the attributes up to the requirements. One should be able to create his own attribute to do that super easily. E.g:

```csharp
public class HasAccessToRoomAttribute : AuthorizeAttribute
{
    public HasAccessToRoomAttribute(int number)
        : base("HasAccessToRoom")
    {
        RoomNumber = number;
    }

    public int RoomNumber { get; }
}
```

```csharp
[HasAccessToRoom(404)]
public IActionResult Test() => Ok();

**Comment by blowdart** [comment]
Ah it's parameterization in the attribute you want. Got it.

To take @PinpointTownes's example, at this point I'd argue we're talking about resource authorization, where the resource is the parameter, a room, with a room number. If you want parameters at runtime that's generally the way to do. The age example of 18,19,20,21 all on different routes is contrived. When we try to design features real world examples are useful in order to justify the dev work.

**Comment by PinpointTownes** [comment]
> To take @PinpointTownes's example, at this point I'd argue we're talking about resource authorization, where the resource is the parameter, a room, with a room number.

Resource authorization means you can't use MVC's built-in authorization stuff and have to rely on manual calls to the authorization service, which is not really ideal.

Really, what we need is a way to do...

```csharp
var result = await _authorizationService.AuthorizeAsync(User, resource: [whatever], requirement: new RoomAccessRequirement
{
    RoomNumber = 404 // the value taken from the attribute.
});
```

... using simple attributes.

**Comment by HaoK** [comment]
Yeah we added IPolicyProvider to at least enable using the policyName as the parameter, so you can to implement a policy provider that resolves Policy="HasAccessToRoom:404", but improving this is on the list of authZ improvements, maybe for 2.2

**Comment by PinpointTownes** [comment]
> so you can to implement a policy provider that resolves Policy="HasAccessToRoom:404"

This is exactly the option demonstrated by @mjrousos in his sample (and the one I personally use). But you can admit it's not ideal as you have to rely on suboptimal string parsing.

**Comment by HaoK** [comment]
Yeah he linked to the sample that drove the feature so yeah :)  I'll create an auth sample for this scenario since its a common ask and pretty hard to discover... its on the list todo, I'll make sure its on the AuthZ roadmap

**Comment by PinpointTownes** [comment]
Good news.

**Comment by HaoK** [comment]
@mjrousos since you already created a sample for this, do you want to work with me in a PR to have it included in AuthSamples? https://github.com/aspnet/AuthSamples/tree/dev/samples

**Comment by mjrousos** [comment]
Sure, that sounds good.

**Comment by HaoK** [comment]
Awesome, do you want to just open a PR copying your sample over to start. then I can point you to the small tweaks we need to make it build on dev, or I can make those changes myself

**Comment by mjrousos** [comment]
Sure. Sounds like a plan.

**Comment by mjrousos** [comment]
@HaoK, I've opened a PR (aspnet/AuthSamples#31) to add my sample to the auth samples repo.

**Comment by blowdart** [comment]
By using attributes parameters you're putting a magic string in there anyway :)

**Comment by MrZarq** [comment]
The policy generator works if you only have one parameter with a limited range. But what if you have multiple? Or, as is my case, a list of values. I want to do something like [HasRight A | C | F], and in some other place [HasRight X | Y]. I can't add a policy for every possible combination of values of my enum now, can I? How do you propose I solve this?

**Comment by HaoK** [comment]
https://github.com/aspnet/AuthSamples/tree/dev/samples/CustomPolicyProvider

**Comment by OneCyrus** [comment]
we have the same issue as @MrZarq. The solution from @HaoK with string concatenation only works with one parameter (and still limited)

**Comment by MrZarq** [comment]
@OneCyrus, we have resolved this problem using the solution from this StackOverflow answer: https://stackoverflow.com/a/40824351/812483

It works, but I have my questions regarding the efficiency seeing as how it uses reflection.

**Comment by OneCyrus** [comment]
thanks! i'll have a look
